### ISSUE TITLE
Failed Tests on M1 MacBook Air - macOS 12.3.1

### ISSUE THREAD

**Issue by theandrew168** [benhoyt/goawk#108]
Hey Ben,

I love this project and all of the blogging that you do related to it. I noticed that a couple of tests are failing on my M1 MacBook Air.

How I run the tests:
```
go build
go test -count=1 ./...
```

Results:
```
ok  	github.com/benhoyt/goawk	1.791s
?   	github.com/benhoyt/goawk/awkgo	[no test files]
ok  	github.com/benhoyt/goawk/internal/ast	0.156s
ok  	github.com/benhoyt/goawk/internal/compiler	0.120s
--- FAIL: TestInterp (1.96s)
    --- FAIL: TestInterp/BEGIN_{_print_sprintf("%u",_-1)_} (0.00s)
        interp_test.go:947: expected "18446744073709551615\n", got "0\n"
--- FAIL: TestNative (0.00s)
    --- FAIL: TestNative/_BEGIN_{__print__print_bool(),_bool(0),_bool(1),_bool(""),_bool("0"),_ (0.00s)
        interp_test.go:947: expected "\n0 0 1 0 1 1\n0 42 -5 3 -3\n0 42 -5 127 -128 -1 0\n0 42 -5 32767 -32768 -1 0\n0 42 -5 2147483647 -2147483648 -2147483648 -2147483648\n0 42 -5 2147483647000 -2147483647000\n0 42 1.84467e+19\n0 42 251 127 128 255 0\n0 42 65535 65535 0\n0 42 4294967295 4294967295 0\n0 42 1.84467e+19 4294967296 2147483647000\n..Foo bar.1234\n..Foo bar.1234\n", got "\n0 0 1 0 1 1\n0 42 -5 3 -3\n0 42 -5 127 -128 -1 0\n0 42 -5 32767 -32768 -1 0\n0 42 -5 2147483647 2147483647 2147483647 2147483647\n0 42 -5 2147483647000 -2147483647000\n0 42 0\n0 42 251 127 128 255 0\n0 42 65535 65535 0\n0 42 4294967295 4294967295 0\n0 42 0 4294967296 2147483647000\n..Foo bar.1234\n..Foo bar.1234\n"
FAIL
FAIL	github.com/benhoyt/goawk/interp	2.270s
ok  	github.com/benhoyt/goawk/lexer	0.073s
ok  	github.com/benhoyt/goawk/parser	0.290s
?   	github.com/benhoyt/goawk/scripts/csvbench/count	[no test files]
?   	github.com/benhoyt/goawk/scripts/csvbench/write	[no test files]
FAIL
```

Both failures seem to be related to handling that large integer value: `18446744073709551615`. Perhaps Go handles large integers differently on ARM devices? Both systems (your Linux machine and my laptop) are 64-bit so we have that in common.

Let me know if I'm running into some user error here or if there is anything I can do to gather more info.

Thanks again for the great work!

**Comment by benhoyt** [comment]
This is very interesting -- thanks for the report! Can you please let me know what output the following program gives on your M1 Mac, and the Go version you're running? ([Playground link.](https://go.dev/play/p/joJA_6J8dk0)) This is basically what the equivalent code in that first GoAWK test is doing.

```go
package main

import "fmt"

func main() {
	f := -1.0
	fmt.Printf("%d", uint(f))
	// Output: 18446744073709551615
}
```

And what about [this one](https://go.dev/play/p/umbvqorpvxo) (equivalent to the int32 failure in the second test):

```go
package main

import "fmt"

func main() {
	f := 2147483648
	i32 := int32(f)
	f2 := float64(int64(i32))
	fmt.Printf("%d", int(f2))
	// Output: -2147483648
}
```

It's possible it's a bug in the Go compiler for M1. I wouldn't think that conversions like this -- even at the boundaries -- should be inconsistent across architectures.

Unfortunately GitHub Actions [does not currently support M1](https://github.com/actions/runner/issues/805), but I'll add that to CI as soon as it does.

**Comment by benhoyt** [comment]
Oh, it looks like these kinds of floating point conversions *are* specified as implementation-dependent [in the spec](https://go.dev/ref/spec#Conversions) under "Conversions between numeric types"

> In all non-constant conversions involving floating-point or complex values, if the result type cannot represent the value the conversion succeeds but the result value is implementation-dependent.

In that case I think I'll just remove those tests so we're not testing implementation-dependent behaviour. I'll do that in the next day or two and release a patch version.

See also:

* https://github.com/golang/go/issues/19405
* https://github.com/golang/go/issues/29468

**Comment by theandrew168** [comment]
Nice find! Good to know about that implementation-dependent behavior in case I ever run into it. As expected, the first example you gave printed `0` while the second printed the expected `-2147483648`.

Removing those specific tests makes sense as the best way to remedy this. Thanks again for all the blogs and code!

**Comment by benhoyt** [comment]
No problem. And I've released v1.17.1 with this fix: https://github.com/benhoyt/goawk/releases/tag/v1.17.1 (I've also updated the `make_binaries.sh` script to add the darwin arm64 build for Mac M1 and added that binary to the new release).
